massa-models/src/denunciation.rs

```diff
@@ -714,6 +731,203 @@ impl Deserializer<Denunciation> for DenunciationDeserializer {
+            .map_err(|_| std::fmt::Error)?;
+        bytes.extend(self.0.to_bytes());
+        write!(
+            f,
+            "{}{}",
+            DENUNCIATION_ID_PREFIX,
+            bs58::encode(bytes).with_check().into_string()
+        )
+    }
+}
+
+impl std::fmt::Debug for DenunciationId {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl DenunciationId {
+    /// denunciation id to bytes
+    pub fn to_bytes(&self) -> &[u8; DENUNCIATION_ID_SIZE_BYTES] {
+        self.0.to_bytes()
+    }
+
+    /// denunciation id into bytes
+    pub fn into_bytes(self) -> [u8; DENUNCIATION_ID_SIZE_BYTES] {
+        self.0.into_bytes()
+    }
+
+    /// denunciation id from bytes
+    pub fn from_bytes(data: &[u8; DENUNCIATION_ID_SIZE_BYTES]) -> DenunciationId {
+        DenunciationId(Hash::from_bytes(data))
+    }
+}
+
+impl From<&Denunciation> for DenunciationId {
+    fn from(value: &Denunciation) -> Self {
+        match value {
+            Denunciation::Endorsement(endo_de) => {
+                let mut to_hash = Vec::with_capacity(
+                    std::mem::size_of::<u32>() + PUBLIC_KEY_SIZE_BYTES + SLOT_KEY_SIZE,
+                );
+                to_hash.extend(endo_de.public_key.to_bytes());
+                to_hash.extend(endo_de.slot.to_bytes_key());
+                to_hash.extend(endo_de.index.to_le_bytes());
+                DenunciationId(Hash::compute_from(&to_hash))
+            }
+            Denunciation::BlockHeader(blkh_de) => {
+                let mut to_hash = Vec::new();
+                to_hash.extend(blkh_de.public_key.to_bytes());
+                to_hash.extend(blkh_de.slot.to_bytes_key());
+                DenunciationId(Hash::compute_from(&to_hash))
+            }
+        }
+    }
+}
+
+// End Denunciation Id
+
+// Denunciation interest
+
+/// DenunciationInterest variant for endorsement
+#[allow(missing_docs)]
+#[derive(Debug, Clone)]
+pub struct EndorsementDenunciationInterest {
+    pub public_key: PublicKey,
+    pub slot: Slot,
+    pub index: u32,
+    hash: Hash,
+    signature: Signature,
+}
+
+/// DenunciationInterest variant for block header
+#[allow(missing_docs)]
+#[derive(Debug, Clone)]
+pub struct BlockHeaderDenunciationInterest {
+    pub public_key: PublicKey,
+    pub slot: Slot,
+    hash: Hash,
+    signature: Signature,
+}
+
+/// Lightweight data for Denunciation creation
+/// (avoid storing heavyweight secured header or secure share endorsement)
+#[allow(missing_docs)]
+#[derive(Debug, Clone)]
+pub enum DenunciationInterest {
+    Endorsement(EndorsementDenunciationInterest),
+    BlockHeader(BlockHeaderDenunciationInterest),
+}
+
+impl TryFrom<&SecureShareEndorsement> for DenunciationInterest {
+    type Error = DenunciationError;
+
+    fn try_from(value: &SecureShareEndorsement) -> Result<Self, Self::Error> {
+        let hash = EndorsementDenunciation::compute_content_hash(&value.content)?;
```

Comment:
It's not necessary to reserialize everything compute the hash, manage potential errors etc...
SecureShare objects store their hash in their ID, just recover it from there like we do for signature verification.

Reply:
This is not the full hash with public key / denunciation data / hash(content) but only this hash(content) so we need to recompute it here...
Otherwise the test that check Denunciation made from 2 block headers versus 2 DenunciationInterest will fail...

Reply:
kinda hurts to have to reserialize the whole data strucutre here...  maybe we should think of a way to store this hash somewhere ?

Reply:
I've added a TODO for this

Reply:
I proposed a correction here: https://github.com/massalabs/massa/pull/3755

Reply:
ok, is there a followup issue with all the TODOs ? because if we just leave them in the code we might forget them

Reply:
I've crated an issue here: https://github.com/massalabs/massa/issues/3758 :)